src/frontend/src/optimizer/plan_node/stream_sink.rs

```diff
@@ -256,38 +273,41 @@ impl StreamSink {
         format_desc: Option<&SinkFormatDesc>,
     ) -> Result<SinkType> {
         let frontend_derived_append_only = input_append_only;
-        let (user_defined_append_only, user_force_append_only, syntax_legacy) = match format_desc {
-            Some(f) => (
-                f.format == SinkFormat::AppendOnly,
-                Self::is_user_force_append_only(&WithOptions::from_inner(f.options.clone()))?,
-                false,
-            ),
-            None => (
-                Self::is_user_defined_append_only(properties)?,
-                Self::is_user_force_append_only(properties)?,
-                true,
-            ),
-        };
+        use UserDefinedAppendOnly::*;
+        let (user_defined_append_only, user_force_append_only, from_format_clause) =
+            match format_desc {
+                Some(f) => (
+                    UserDefinedAppendOnly::from(&f.format),
+                    Self::is_user_force_append_only(&WithOptions::from_inner(f.options.clone()))?
+                        || Self::is_user_force_append_only(properties)?,
+                    false,
+                ),
+                None => (
+                    Self::user_defined_append_only(properties)?,
+                    Self::is_user_force_append_only(properties)?,
+                    true,
+                ),
+            };
 
         match (
             frontend_derived_append_only,
             user_defined_append_only,
             user_force_append_only,
         ) {
-            (true, true, _) => Ok(SinkType::AppendOnly),
-            (false, true, true) => Ok(SinkType::ForceAppendOnly),
-            (_, false, false) => Ok(SinkType::Upsert),
-            (false, true, false) => {
+            (true, DefinedAppendOnly|NotDefined, _) => Ok(SinkType::AppendOnly),
```

Comment:
Thanks for your clarification. If we do not maintain the compatibility for those errored statements, the only change is this one.

> this is a backward incompatible change: NotDefined was considered false before this PR, even when frontend derives it is actually append only.

Reply:
Thanks for your clarification. If we do not maintain the compatibility for those errored statements, the only change is this one.

> this is a backward incompatible change: NotDefined was considered false before this PR, even when frontend derives it is actually append only.

Reply:
> If we do not maintain the compatibility for those errored statements

In my opinion, it is okay now to move from err to non-err in this PR. However, if we want to move from non-err to err again in the future, it would be a breaking change. This is what I mean by *making the backward compatibility promise harder to maintain*.